# Cancellation Model Framing

This notebook defines the machine learning target, removes leakage-prone columns, selects pre-arrival features, and creates the train/test split for cancellation prediction.

In [ ]:
# ADD IMPORTS AND LOAD DATA

import pandas as pd
from sklearn.model_selection import train_test_split

df = pd.read_csv("../data/processed/hotel_bookings_cleaned.csv")

df.shape

(119390, 36)

In [2]:
# DEFINE TARGET

target = "is_canceled"

df[target].value_counts(normalize=True) * 100

is_canceled
0    62.958372
1    37.041628
Name: proportion, dtype: float64

In [3]:
# REMOVE LEAKAGE COLUMNS

leakage_columns = [
    "reservation_status",
    "reservation_status_date"
]

df_model = df.drop(columns=leakage_columns, errors="ignore")

df_model.shape

(119390, 34)

In [ ]:
# BASELINE FEATURES

selected_features = [
    "hotel",
    "lead_time",
    "arrival_date_month",
    "stays_in_weekend_nights",
    "stays_in_week_nights",
    "adults",
    "children",
    "babies",
    "meal",
    "country",
    "market_segment",
    "distribution_channel",
    "is_repeated_guest",
    "previous_cancellations",
    "previous_bookings_not_canceled",
    "reserved_room_type",
    "booking_changes",
    "deposit_type",
    "agent",
    "company",
    "days_in_waiting_list",
    "customer_type",
    "adr",
    "required_car_parking_spaces",
    "total_of_special_requests",
    "total_nights",
    "total_guests",
    "arrival_date_num",
    "stay_type"
]

X = df_model[selected_features]
y = df_model[target]

X.shape, y.shape

((119390, 29), (119390,))

In [6]:
# CREATE TRAIL/TEST SPLIT

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

X_train.shape, X_test.shape, y_train.shape, y_test.shape

((95512, 29), (23878, 29), (95512,), (23878,))

In [7]:
# CHECK TARGET BALANCE

print("Training target distribution:")
print(y_train.value_counts(normalize=True) * 100)

print("\nTesting target distribution:")
print(y_test.value_counts(normalize=True) * 100)

Training target distribution:
is_canceled
0    62.958581
1    37.041419
Name: proportion, dtype: float64

Testing target distribution:
is_canceled
0    62.957534
1    37.042466
Name: proportion, dtype: float64


In [8]:
# SAVE BASELINE MODELLING DATASETS

train_data = X_train.copy()
train_data[target] = y_train

test_data = X_test.copy()
test_data[target] = y_test

train_data.to_csv("../data/processed/model_train.csv", index=False)
test_data.to_csv("../data/processed/model_test.csv", index=False)

print("Model train and test datasets saved successfully.")

Model train and test datasets saved successfully.


# Day 11 Summary

## Model Framing Completed

### Target Variable
- is_canceled

### Data Leakage Prevention
Removed:
- reservation_status
- reservation_status_date

### Baseline Features
Selected 29 booking-related features available before cancellation occurs.

### Train-Test Split
- Training Set: 95,512 rows
- Testing Set: 23,878 rows
- Split Ratio: 80:20

### Class Distribution
Training:
- Not Cancelled: 62.96%
- Cancelled: 37.04%

Testing:
- Not Cancelled: 62.96%
- Cancelled: 37.04%

### Outcome
Prepared clean train and test datasets for machine learning model development.